spec/consensus/fork-detection.md

```diff
@@ -0,0 +1,45 @@
+
+processing misbehavior on the main chain (if possible).
+
+## Detecting forks
+
+A fork is defined as a deviation (of list of blocks) from the main chain. In this document we will assume that there exists a main chain C1 (sequence of blocks) and an honest full node F that is synced to the chain C1. Then a fork corresponds to an existence of a commit for block B for height h that is different from a block at height h on the chain C1. Furthermore, a commit for the block B is a valid fork if it contains at least a single valid signature by a validator from the validator set at height h at chain C1. This definition of valid fork is needed to differentiate forks that can provide material for punishment mechanisms (slashing) from bogus data that looks like a valid commit but is signed by processes that are not tracked by the system (and therefore can't be punished).
+
+As explained in lite-client document a lite client establish a connection with a full node and it verifies application states of interest by downloading and verifying corresponding block headers. To be able to detect a fork, we assume an additional module (fork detection module) that runs in parallel with the core lite client module, and whose role is fork detection. The fork detection module maintains a list of peers (full nodes) and it periodically establishes a connection to some peer (called witness) and verify correctness of headers downloaded by main module (by cross-checking with headers downloaded from witnesses). In case there are two different headers for the same height signed by overlapping validator sets, then a client downloads corresponding commit and creates a proof of valid fork.
+
+## Identifying faulty validators
+
+Once an honest lite client detects a valid fork it needs to submit it to an honest full node for further processing so that faulty validators can be detected and punished. For the purpose of this document we assume that a lite client is able to submit a proof of fork to an honest full node in reasonable time (how we can ensure that will be addressed in a separate document).
+
+For the purpose of this document we assume existence of the following information for each commit: hash of the block, height, round, validator set and a set of signatures by the processes in the given validator set. An honest full node starts the procedure that detects misbehaving validators from the following information: for some height h, there are two conflicting commits C1 and C2, where C1 is a commit from the main chain. The detection procedure works like this:
+
+if there are processes from the set signers(C2) that are not part of C1.valset, they misbehaved as they are signing protocol messages in heights they are not validators. In this case commit C2 is a self-contained evidence of misbehavior for those processes that can be simply verified by every honest validators without additional information.
```

Comment:
That's why light client can (only) do header comparisons, and in case it detects a fork it needs to send this information to a correct full node for further processing. You are right that both primary and witnesses can be faulty. We need an additional mechanism that guarantees that eventually light client will submit fork to a correct full node. To ensure this we probably need light client to manage an address book (with particular properties) and to keep submitting evidence to a random set from the address book long enough, so a correct node is eventually reached. This part is missing in this document.